core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -983,6 +1021,38 @@ private void ensureMutable() {
         }
     }
 
+    public Iterator<ShardRouting> nodeInterleavedShardIterator() {
+        final Queue<Iterator<ShardRouting>> queue = new ArrayDeque<>();
+        for (Map.Entry<String, RoutingNode> entry : nodesToShards.entrySet()) {
+            queue.add(entry.getValue().copyShards().iterator());
+        }
+        return new Iterator<ShardRouting>() {
```

Comment:
`Iterators#concat` will give us all the shards of the first node, then all the shards of the second node, etc. The iteration order here is different: We take a shard from each node in succession. I've added docs describing the iteration order.